### ISSUE TITLE
Future of this addon

### ISSUE THREAD

**Issue by jembezmamy** [webark/ember-cli-styles#1]
Hey @webark, I've landed here from https://github.com/ebryn/ember-component-css/issues/355 so you probably know what I'm going to say. :)

I'm trying to migrate my app to Embroider and ember-component-css is currently the biggest obstacle. So it would be great if we could make this addon work and publish at least a basic version of it.

So my question is: how can I help? :) What needs to be done here to make this addon work with Glimmer and Embroider?

Also, I tried to install the most recent version and there are some issues with `ember-cli-styles-colocation@*` dependency, could you please fix that? I'm not sure what was the intention here. :) I tried symlinking https://github.com/webark/ember-cli-styles-collocation but without any luck.

**Comment by webark** [comment]
@jembezmamy if you were able to get a sample app setup with embroider and some styles, I'll get it working for you by next week.

**Comment by jembezmamy** [comment]
Wow, thanks, that would be great! I'll prepare the sample app and let you know.

**Comment by burritoIand** [comment]
thank you both, have enjoyed using ember-component-css and would like to migrate to this once it's ready so I can update my app

**Comment by webark** [comment]
oh gosh. I just opened this up and my latest changes were not pushed.

**Comment by webark** [comment]
I pushed those, but it seems that the tests are not running. They were running before. I'll look into this and see what's happening.. It seems something screwy is going on.. (this is what happens when you come back after a year of not looking at it 😂 )

**Comment by webark** [comment]
Ok. I pushed everything and the tests for "ember-cli-styles-namespace" are passing. So should be able to look there how to use this. I haven't published beta versions of these packages yet as an FYI

**Comment by jembezmamy** [comment]
@webark Ha ha, great, thanks. 😄  I'm going to work on the embroider demo next week, currently I'm on a short vacation.

**Comment by webark** [comment]
haha!! I'm on vacation too.. First "two night" vacation i've had with my wife since our daughter was born 6 YEARS ago!! I'm sorry that i've dragged this out.. I'm really interested in this working with embroider, and have some time in the next couple of months to spend on it. If someone can get an embroider app setup with this, i'll be able to make sure it's working.

**Comment by burritoIand** [comment]
I might take a stab at upgrading soon.. I'm on 3.20 still. Big thanks to you both for working on this.

Random question: how hard would it be to add option to minimize all classes e.g. "__3824j__my-class--foo-123" -> "xy123"

**Comment by webark** [comment]
there's that option in the current iteration, I forget if I carried that over to this "namespace" method.

**Comment by boris-petrov** [comment]
@webark - in addition to adding Embroider support (which would be nice, of course) I believe a couple of more things are more important - namely Ember 4.0 support, tagless/template-only/Glimmer components support and co-location.

Embroider, even if released soon, won't replace `ember-cli` which will still be supported for probably years to come. However, Ember 4 is coming very soon and `ember-component-css` won't work on it because of some [deprecation warnings](https://github.com/ebryn/ember-component-css/issues/365). I believe that is the most important thing for `ember-cli-styles` to fix - so that this addon is usable on the latest Ember version.

Tagless components support would also be nice but there is a workaround for that right now - using `stylesNamespace`. So that could wait.

Co-location also would be very nice but that could also wait.

If you need example apps for any/all of these issues, just tell me and I'll prepare them immediately. :)

Thank you for taking the time to work on this! We love `ember-component-css` and would like to continue to use it going forward with the latest Ember versions. :)

**Comment by webark** [comment]
This rewrite handles these cases the best way I have found to satisfy tagless components. Currently it wraps each temple with a `let` block providing access to the `styleNamespace` property to be used in the template. This requires the developer to specifically add this class to those tagless components, but to keep to the same spirit of how this addon handles its notion of "namespacing" styles (i.e. `.namespace SELECTOR`) If there are other methods, i an open to explore those.

Co-location (as I understand it) is also supported. Single file components are not supported, but those are experimental, and support could be added later.

I have not attempted to use it in an embroider setup, but the new mechanism should carry over. 

The main thing I need to work on to get this released, is the legacy support (which i'd be open to releasing these addons without that) and some real wold flexing and vetting of this solution. (which I have not done as I am no longer working on ember apps, or much in front end apps in general and am, sadly 😩 in react world when I do now).

If a subset of you all are open to helping me vet this new approach, I will commit to offering assistance and continued development as I can in my spare time. And if people start to contribute, give privileges of shared ownership of).

Does that answer your questions? If people are open to a zoom sometime, I'd be open to that as well.

**Comment by boris-petrov** [comment]
Thanks for the comment!

I believe that wrapping each template in a `let` block is fine...ish. Right now I'm using a helper which uses `podNames` from `ember-component-css/pod-names` and pass it the component's name. That is definitely not ideal but perhaps better in some way - as the template is not modified in any way by the plugin. Not sure though - perhaps the `let` approach is also OK.

As for legacy support - I think that could be skipped entirely given that you don't have all the time in the world. You can always release a new major version and tell people they should change this and that in order to continue to use the addon. It's totally fine to break backwards-compatibility (even greatly) in order to be able to support a package. The other way means we don't get anywhere because you're trying to implement/support features that you don't have the time for. So if you think the current state of affairs works - you can try releasing a beta, explain what is changed for users of the plugin, and wait for bug reports. :) It's easier for people to change a few things here and there on how they use the addon than a maintainer to support a million old things.

I didn't understand about Ember 4 support - are the deprecated things fixed and is Ember 4 supported?

**Comment by webark** [comment]
Yes to ember 4 support. It should be fully supported with this addon as long as the breaking changes are addressed.

**Comment by webark** [comment]
And ok. That's a good point about legacy support. I will tie up some loose ends and get a beta release out of these.

**Comment by jembezmamy** [comment]
@webark I've pulled the newest version of this repo and  I still have hard time installing this locally (yarn install fails because of missing ember-cli-styles-colocation package, I suppose it's a git submodule but I can't pull this cause it's missing url or something). So maybe I'll wait for the beta release with further experiments. :D

**Comment by webark** [comment]
I recently fixed that. You should be able to just yarn link it.

**Comment by webark** [comment]
wait what? You are right.. I thought I just pushed that.

**Comment by webark** [comment]
OOOKKK... Now there we go. Sorry about that @jembezmamy

**Comment by joelcox** [comment]
What would be the best way to help/test this add-on? I don't have any practical experience with developing Ember addons, but sure could give it a try if that helps. What is the best way to test-drive this? Pull the Lerna repository and individually link all different packages into a project?

**Comment by webark** [comment]
Ok. I just rewrote the git history of this cause it was in a messed upstate. And cleaned up some loose edges. I have a couple of style things to get the lint passed for the tests, then I will release a v1 beta. I'll also add some basic usage so people can see how things have changed.

The next parts after that would be adding a migration plan (which will be minimal) and then finishing the addon for the legacy support (won't be terrible).

I know this is all way too late.. But better late than never. Sorry for the burden having an outdated package has been :/

**Comment by boris-petrov** [comment]
Absolutely no problem! Thank you for still taking the time to work on this! Also, Ember 4 is still not out so it’s definitely not too late. :) I’ll await the beta release and will try it out as I want to get rid of the deprecation warnings (and soon update to Ember 4).

**Comment by webark** [comment]
So it looks ike ebroider doesn't support normal pods, and only supports colocation. So I'm going to update the tests to that, then I should be able to release the beta.

**Comment by webark** [comment]
OK. Just pushed a change to switch from modifiers to helpers based on feedback from @SergeAstapov in which fastboot doesn't work with modifiers. (was brought up in https://github.com/ebryn/ember-component-css/issues/335)

The last things to do are...

- [ ] Embroider support: (they pass in a different broccoli tree to the css preprocessor which doesn't include js files. we handle this in addons, but non embroider doesn't include them for addons an apps.)
- [ ] Pre 3.24 is having issues with the ast and the argument that I'm adding to the helper. When I manually add it using the same syntax it works, so I'm a little stumped on this one, but will circle back around to it after the emborider fix
- [ ] Adding in the basic documentation on how to use the new helper and handle the other updates.

**Comment by webark** [comment]
I also have never released something with lerna.. so if someone knows the command to trigger an alpha release, I will run that.

**Comment by webark** [comment]
release it looks fancy! Maybe i'll try that out.

**Comment by boris-petrov** [comment]
@webark - https://github.com/boris-petrov/ember-component-css-example

Clone the repo, `npm install`, `ember serve`. The relevant code is in `app/controllers/application.js`:

```js
getZone('<div class={{style-namespace "some-component"}}><a href="#">aaaaa</a></div>')
```

This generates, at runtime, the following element:

```html
<div class="__some-component__45cbc"><a href="#">aaaaa</a></div>
```

And the correct CSS is applied.

Please ask if there is something you don't understand.

**Comment by boris-petrov** [comment]
@webark - I don't want to sound pushy but are there news on the beta release? :smiley: Ember 4, as I understand, is coming in a week or two so it would be nice to migrate from `ember-component-css` as soon as possible. Thank you! :)

**Comment by webark** [comment]
@boris-petrov ok. I published what I had so far. 

here's an example.

https://github.com/webark/ember-cli-styles-example

right now you might have to add both
```    "ember-cli-styles-colocation": "^1.0.0-alpha.1",
    "ember-cli-styles-namespace": "^1.0.0-alpha.4",
```

I haven't looked into your example yet. I will do that this week.

**Comment by boris-petrov** [comment]
@webark - thank you very much!

I believe you meant that both `ember-cli-styles-namespace` and `ember-cli-styles-preprocessor` must be added, because `ember-cli-styles-colocation` is a dependency of `ember-cli-styles-namespace` while `ember-cli-styles-preprocessor` isn't.

Here's what I did:

1) I moved `app/styles/components` to `app/components` so that styles are co-located.

2) I removed my own implementation of `helpers/style-namespace.ts`.

3) In `app/styles/app.scss` I replaced `@import 'pod-styles';` with `@import 'ember-styles.css';`.

4) I removed `ENV['ember-component-css'] = { classicStyleDir: 'components' };` from `config/environment.js`.

Issues I have:

1) Running `ember serve` lead to an SCSS error for starters - in my `app.scss` I'm using an SCSS variable that is defined in one of the component SCSS files that obviously previously was made available to SCSS via `@import 'pod-styles';` but now isn't for some reason. Any ideas why?

2) Fixing this issue `ember serve` started fine but going to the website reveals that CSS is missing and in the file where it should be there are 5 lines of `/* Empty File. If know how to conditionaly run broccoli plugins, please reach out. Thanks. */`, I guess coming from [here](https://github.com/webark/ember-cli-styles/blob/ember-cli-styles-preprocessor%401.0.0-alpha.2/packages/preprocessor/lib/preprocessor.js#L62). Is somehow SCSS not supported?

3) A major issue that I saw - the function `add-component-style-namespace` [here](https://github.com/webark/ember-cli-styles/blob/ember-cli-styles-preprocessor%401.0.0-alpha.2/packages/namespace/addon/utils/add-component-style-namespace.js#L8) looks-up all components on page-load (or at least the ones that have styles I guess?). This causes Ember to create an instance of this component and sends the `init` event - which causes observers on `init` to be triggered, `init` methods to be invoked and all horrible kinds of things happen because of that. This issue is a show-stopper. I'm not sure if Ember allows looking-up a component without it being instantiated.

Thanks again for your time! I'll appreciate any help on these issues.

**Comment by webark** [comment]
@boris-petrov you don't need the preprocessor.. you can use your own. You need both the collocation and the namespace cause they both have preprocessors that need to run, and the collocation one wasn't running unless O added it specifically.

**Comment by webark** [comment]
The "empty file" comment is from the preprocrssor. It creates an empty style file if it doesn't exists, wasn't sure how to handle it otherwise. If you know i'm all ears.

**Comment by webark** [comment]
> A major issue that I saw - the function add-component-style-namespace here looks-up all components on page-load (or at least the ones that have styles I guess?). This causes Ember to create an instance of this component and sends the init event - which causes observers on init to be triggered, init methods to be invoked and all horrible kinds of things happen because of that. This issue is a show-stopper. I'm not sure if Ember allows looking-up a component without it being instantiated.

The need here is to auto add the namespace to the component. Do you know of a way that could be handled more efficiently?

**Comment by webark** [comment]
@boris-petrov I updated https://github.com/webark/ember-cli-styles-example with an example using scss variables.

**Comment by webark** [comment]
and @boris-petrov I don't have permission to your repo, but was able to achive the support by doing

```
diff --git a/app/controllers/application.js b/app/controllers/application.js
index d7ee272..292e76b 100644
--- a/app/controllers/application.js
+++ b/app/controllers/application.js
@@ -1,5 +1,6 @@
 import Controller from '@ember/controller';
 import { compileTemplate } from '@ember/template-compilation';
+import styleManifest from 'ember-component-css-example/components/some-component.css';
 
 function registerTemplateOnlyGlimmerComponent(moduleName, getTemplate) {
   define(moduleName, ['exports'], function (_exports) {
@@ -19,5 +20,5 @@ function getZone(template) {
 }
 
 export default class ApplicationController extends Controller {
-  zone = getZone('<div class={{style-namespace "some-component"}}><a href="#">aaaaa</a></div>')
+  zone = getZone(`<div class=${styleManifest.styleNamespace}><a href="#">aaaaa</a></div>`)
 }
```

not sure why I coudln't

```
import { styleNamespace } from 'ember-component-css-example/components/some-component.css'
```

I'll have to look into that.

**Comment by boris-petrov** [comment]
@webark - thanks for the help! I used `@import 'ember-styles.css';` instead of `@import 'ember-styles.scss';` in my `app.scss` and that's why issue number 1 appeared. Changing it, however, leads to another trouble:

```
File to import not found or unreadable
  ╷
1 │ @import "my-app/components/some-component.scss";
  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

That's because I'm using `ember-cli-postcss`. I've added a [commit](https://github.com/boris-petrov/ember-cli-styles-example/commit/16e08045d171bad9f9e0134c1d4c9adaf2c459e0) on your example repo to showcase the issue. Just run `ember serve` and you'll see it. What are we to do with that?

**Comment by webark** [comment]
@boris-petrov so. It looks like I'll have to open up the ability to customize the manifest generation sooner than later.. changing it to https://github.com/webark/ember-cli-styles/blob/87b40db1d1d81fc99efe7704b74de61e4640cc72/packages/colocation/lib/colocate.js#L24-L29 made it able to find the file. I don't know if you can get that resolution path othersize.

**Comment by boris-petrov** [comment]
@webark - that does seem to make it compile, yes. That leads to another issue, however:

```
Top-level selectors may not contain the parent selector "&".
  ╷
1 │ & {
  │ ^^
  ╵
```
There seems to be some change in handling of SCSS files because that wasn't the case with `ember-component-css`. Ideas about that? :) Note that this is again using `ember-cli-postcss`.

**Comment by webark** [comment]
so it should be using the same handling of that. 

https://github.com/webark/ember-cli-styles/blob/87b40db1d1d81fc99efe7704b74de61e4640cc72/packages/namespace/lib/preprocess-class-names.js#L35-L39

https://github.com/ebryn/ember-component-css/blob/a39a8137844527e22eb42e0a31523b9972d0a30f/lib/preprocess-class-names.js#L38-L43

**Comment by webark** [comment]
https://github.com/topaxi/postcss-selector-namespace got updated. It might have changed how it handles that? Or maybe it's a plugin ordering issue? 

you could try the `:--component` syntax?

**Comment by boris-petrov** [comment]
Sorry, I'm not sure I follow what you mean. Did you manage to reproduce the problem? Why is there a difference between `ember-component-css` and `ember-cli-styles` in this regard?

**Comment by webark** [comment]
we have always used a postcss plug-in to process the scss files. The package was at v 2 now it's at v3. I'm just wondering if there's a breaking change in that, or if the ordering of the addons/postcss plug-ins is not effected. Another avenue to explore.

**Comment by webark** [comment]
so when I log out https://github.com/webark/ember-cli-styles/blob/87b40db1d1d81fc99efe7704b74de61e4640cc72/packages/namespace/lib/preprocess-class-names.js#L35-L39
the transformation is there. So postcss must be picking up the raw file again.

**Comment by webark** [comment]
I'll have to come back to this.. Spent too much time today sorry.. might not be till next week. Feel free to go diving though the way postcss is loading and parsing the files. I'm not sure why the `ember-cli-styles-preprocessor` is working but this is not.

**Comment by boris-petrov** [comment]
@webark - thanks for digging in some of the problems that I encountered! I'm not sure how much of help I can be with debugging `ember-cli-styles` itself as I don't understand the process pretty much at all. I can help mostly with reproductions of issues.

One thing I noticed is that `ember-cli-styles` uses `postcss-preset-env` which doesn't seem very well supported and for example doesn't work with [PostCSS 8](https://github.com/csstools/postcss-preset-env/issues/191). Other parts of `ember-cli-styles` depend on PostCSS 8 things. Not sure if that's relevant but just an observation.

In any case, thanks again for the time and do let us know when you have any idea on the issue above!

**Comment by webark** [comment]
postcss-preset-env seemed to be the preferred way to handle this back when i first wrote the preprocessor. According to the readme, it still suggests setting it up this way. https://github.com/jeffjewiss/ember-cli-postcss But again, the preprocessor shouldn't be nessicary. It seems like somewhere the resolution is reaching back into the raw file again, rather then pulling the files from the broccoli node. Maybe it's due to how postcss 8 handles its internal caching I'll have to dig in more this weekend or next week.

**Comment by webark** [comment]
@boris-petrov I updatde the packages to be able to pass in custom options. Opened a PR against your example with it working. 

https://github.com/boris-petrov/ember-cli-styles-example/pull/1

I think the issue that you were running into is with how https://github.com/jeffjewiss/ember-cli-postcss was managing the broccoli nodes. I haven't fully debugged what that is, but you are able to use `ember-cli-styles-preprocessor` to get it working.

the customized options I set were..

```javascript
    emberCliStylesOptions: {
      manifestTemplates: {
        scss: '@import "../../<file-path>";',
      },
      preprocessors: {
        scss: {
          broccoliPlugin: require('broccoli-postcss-single'),
          options: {
            map: true,
            extension: 'scss',
            enabled: false,
            parser: require('postcss-scss'),
            plugins: [
              {
                module: require('@csstools/postcss-sass'),
              },
            ],
          },
        },
      },
    },
```

and you can see those in the PR I mentioned to your fork of the example.

**Comment by webark** [comment]
The example uses both a css and scss.. I am using the default config for css preprocessor which uses `postcss-preset-env` but you could overwrite those similarly to how I overwrote the scss preprocessor.

**Comment by webark** [comment]
@boris-petrov let me know if this works for you, and what your next blocker is 🤪

I'll try if I have time to figure out what is wrong with `ember-cli-postcss ` but switching it out for `ember-cli-styles-preprocessor` should be innocuous now that you can fully customize the options it runs.

**Comment by webark** [comment]
I guess it's doing some fliter vs compile stuff that `ember-cli-styles-preprocessor` isnt doing. So i'll still need to look into that and also what's broken with `ember-cli-postcss`

**Comment by boris-petrov** [comment]
@webark - thank you, again, for taking the time to help! I'll check what you've given me in a bit. As for `ember-cli-postcss` - there is also an issue with using `SCSS` in version 7 (even though in my example here I was using version 6) - [issue here](https://github.com/jeffjewiss/ember-cli-postcss/issues/725). If you can figure out what the problem is in `ember-cli-postcss` you'll help a lot of people. :smile:

**Comment by boris-petrov** [comment]
OK, so a couple of quick observations. On my project your solution doesn't work. I get:

```
Class extends value undefined is not a constructor or null
```

When running `ember serve`. The issue is in `node_modules/postcss-scss/lib/nested-declaration.js:3:33` which does:

```js
const { Container } = require('postcss')

class NestedDeclaration extends Container {
```

I believe it's possible to be because of different versions of PostCSS installed. Even in your example there are (but I guess the used one is the correct one and in my project it's the wrong one). E.g. `broccoli-postcss-single` depends on PostCSS v8 while `@csstools/postcss-sass` depends on PostCSS v7. No idea how to fix that...

**Comment by webark** [comment]
@boris-petrov do you have a public example of the project that doesn't work?

**Comment by boris-petrov** [comment]
@webark - I've pushed a [commit](https://github.com/boris-petrov/ember-cli-styles-example/commit/62a70889e49f3a63600740856f0c409e802047ef) showcasing the problem. Note that it's using `yarn` instead of `npm`. That is, you'll have to `rm -rf node_modules && yarn && ember server` to reproduce. When using `npm install` for some reason that doesn't happen...

**Comment by webark** [comment]
@boris-petrov I opened a PR https://github.com/boris-petrov/ember-cli-styles-example/pull/2 against your example. `postcss-scss` just looks like it requires `postcss` 8 to be a dependency. After that, it's working fine. The only other change was adding `ember-page-title` cause the example still needs thatt.

**Comment by boris-petrov** [comment]
@webark - oh well, I can't imagine why adding something to `package.json` (which doesn't even change the lock file as it's a transitive dependency) could fix that issue... but you're right, it does. With that, my project seems to build fine. Great work, thank you!

So now a couple of issues are left.

1) (mandatory) the fact that `ember-cli-styles` looks-up (and hence instantiate) all components with styles at boot time. This is both a performance problem and a logic one - components are supposed to be instantiated (by Ember) only when they're used. For me this **must** be fixed before `ember-cli-styles` could be used in production.

2) (optional) find out what's the issue with `ember-cli-postcss`. This is not strictly needed as there is a [workaround](https://github.com/webark/ember-cli-styles/issues/1#issuecomment-968126487) now. If we're going to stay with it, would you mind telling us how to also add a `filter` section as in `ember-cli-postcss`? Currently your example of `preprocessors: { scss: ...` maps to `compile` I believe?

**Comment by webark** [comment]
@boris-petrov for # 2 yes, this is a "compile" as in it is run during the preprocess step, rather than a "filter" which is just run during the postprocess step.

for # 1 .... 😔 The think is.. manners by which this can be done "automatically" from my knowledge have all closed. I can look back and see if for an `EmberComponent` (which they say will be around for all of ember 4.0) the option of "reopening" is still on the table. Moving forward, I guess glimmer components are the defacto solution, so we shouldn't expect components to be given wrapping elements anyway, and then the need to use the helper will be forced. Perhaps the ember component part could be moved into a legacy support addon.

🤔

**Comment by webark** [comment]
maybe you can inject a computed property into components that would be able to do it.. 🤔

**Comment by boris-petrov** [comment]
@webark - so for issue number 2 - is there a way to also add a `filter` step? As I currently have such in my `ember-cli-postcss` config and would like to migrate that as well.

For number 1 - the thing you do in that code is to set the correct style file to the component? If there is no "automatic" way - why don't you revert to "manual"? That works totally fine right now (even though it's verbose to pass a string as an argument). Also, you could make it accept both a string and a "real" style like in the [example](https://github.com/webark/ember-cli-styles/issues/1#issuecomment-963422524) you gave. In any case that's much better as the way it is right now is unusable. :cry:

**Comment by webark** [comment]
@boris-petrov yea. I think i will add another addon, something like "ember-component-auto-namespace" that will handle this legacy behavior, and then the standard behavior will be the manual process of the helper, or importing the namespace and adding it the the component class.

**Comment by boris-petrov** [comment]
You mean the `ember-component-auto-namespace` addon will handle the NEW behavior? And the default will be the legacy way - manually specifying a string or the imported namespace? Or did I misunderstand you?

**Comment by webark** [comment]
no, that would handle the old way of adding the `styleNamespace` to the "classic componet" (i.e. ember component).

**Comment by boris-petrov** [comment]
Oh, I understand, you mean that the current way of automatically finding all components will be gone and only a manual way will be available (plus this new addon that could automate things for classic components)?

**Comment by webark** [comment]
and even though "Ember Components" are not going away in 4.0... it looks like both injections are off the table https://deprecations.emberjs.com/v3.x/#toc_implicit-injections and reopening https://deprecations.emberjs.com/v3.x/#toc_ember-built-in-components-reopen are off the table (like I had thought). So there is no way to handle this, as the way of looking up all classic components is not viable.. :( At least there is no way that I can see to accomplish this.

**Comment by webark** [comment]
> only a manual way will be available

Correct. And anyone migrating an application will need to go through and add them import and assignment to each and every component class in their application and addons. :(

**Comment by webark** [comment]
Well.. I'll remove that bit of functionality sometime in the next couple of days, and release a new package.

**Comment by boris-petrov** [comment]
Well, for me at least that's totally fine because anyway I'm doing the same right now for Glimmer components (which are the majority of components I have) with `ember-component-css`. I think that's the right way to go.

Thanks! After the release I'll try it out and let you know how it goes. :)

**Comment by webark** [comment]
back before i had landed on the lookup behavior, i was planning on just issuing a code mod. I had written up the babel ast. Maybe If i have the time I could dust that off and put it in a proper ember code mod... Big emphasis on "IF" i have the time.

**Comment by boris-petrov** [comment]
That would be nice but it's not a blocker for users - just more painful. :) I would prefer for all the blockers to be fixed first and then the nice-to-have things to (eventually) follow so that we can upgrade.

**Comment by boris-petrov** [comment]
@webark - any luck removing the auto-discovery functionality and enabling the "manual" version? :)

**Comment by boris-petrov** [comment]
@webark - I just remembered that because there won't be an auto-discovery mechanism for old classic components as in `ember-component-css`, then there should be a way to manually set the class like before:

```ts
import { classNames } from '@ember-decorators/component';
import Component from '@ember/component';
import podNames from 'ember-component-css/pod-names';

@classNames(podNames['some-component'])
class SomeComponent extends Component {
```

**Comment by webark** [comment]
@boris-petrov I pushed the removal of the auto-discovery and also added an example in my example repo for your above example
https://github.com/webark/ember-cli-styles-example/blob/c8fbaec325f6c566037578eee6218e80bf2bc62c/app/components/test-classic.js#L1-L8

```
import { classNames } from '@ember-decorators/component';
import Component from '@ember/component';
import { styleNamespace } from './test-two.scss';

@classNames(styleNamespace)
export default class SomeComponent extends Component {

}
```

**Comment by webark** [comment]
@boris-petrov you would just have to import the namespace from wherever you want to use it. The CSS get scoped (as it does with ecc) during build time. So passing in a random string won't effect the CSS that has already been generated.

Do you have an example of how you are currently doing this with ecc and I can try to replicate that example?

**Comment by boris-petrov** [comment]
@webark - sure, [here you go](https://github.com/boris-petrov/ember-component-css-example). Only look at the third commit. This is the relevant code:

```hbs
<Input @value={{this.namespace}} />

<div class={{style-namespace this.namespace}}>
  <a>Text</a>
</div> 
```

If you write `some-component` in the input field, you'll see the text turn red. Any other value will not make it red. That's because there is a `some-component.css` file which is applied when `this.namespace` == `some-component`. Is that possible with `ember-cli-styles`?

**Comment by webark** [comment]
@boris-petrov https://github.com/boris-petrov/ember-component-css-example/pull/1

Basically, you just have to import the styleNamespace from the style file you want to use the namespace of and apply it where you want to apply it.

**Comment by boris-petrov** [comment]
@webark - sorry, I shouldn't have used `namespace` as a variable name, it's confusing. I force-pushed the commit (and edited my previous comment) - I've now used `foobar`.

You've misunderstood my intention. The idea is that I want to apply `style-namespace` with DIFFERENT values - here depending on user-input. If the user types "some-component" in the input field - some CSS is going to be applied. If the user types "other-component" - different CSS is going to be applied.

Run the app, type `some-component` in the input and see how the text below changes. Then, type `other-component` - you'll see different CSS applied. Sorry, maybe I sound a bit confusing, I'm not sure how to better explain. :)

**Comment by webark** [comment]
hmm.. I'll look at it another time. You'll either need to grab the list by generating it yourself through an import manifest of options, or you can look at the "hack" I did for route styles 

https://github.com/webark/ember-cli-styles/blob/4ead8b009e1905a6d8d3469edc7595c57e97b448/packages/namespace/addon/utils/process-style-type.js#L16-L25

**Comment by webark** [comment]
there can be no concept of the `pod-names` though. This is one of the main causes for needing to rewrite a lot of this addon. We can no longer reliably generate an aggregated list of these throughout all of the app styles and addon styles, and with embroider will be impossible.

You can generate one yourself through a self-managed import manifest, or you have to use the `'container-debug-adapter:main'` which I'm unsure what that support will be long-term.

**Comment by boris-petrov** [comment]
@webark - any luck figuring out how to do what I requested above? To help a bit - do you have any ideas how [ember-css-modules](https://github.com/salsify/ember-css-modules) works? It does something similar to `ember-cli-styles` and I believe it does have some concept like `pod-names`. Perhaps you can take some inspiration from there? :)

**Comment by webark** [comment]
@boris-petrov from my understanding at least, packages need to be static, and can not/should not be dynamically modified by other packages. I am not familiar with how ECM works. Do you have a coded link that you could provide?

From my understanding, all your concerns have been addressed, and you will need to maintain your own import manifest of possible components from which you want to dynamically generate and pull class names from.

basically, import all `import { styleNamespace } from './test-two.scss';` from everywhere in your application you want, and then export that in a place you can look up.

**Comment by webark** [comment]
I'm going to close this issue as its primary purpose has been resolved. If you want to open a new issue specifically around a generated manifest that the addon provides, we can pursue that in a new issue. If you have any documentation or code references for that in ECM, that would be appreciated.

Thanks for working through all the issues we did on this one! @boris-petrov your beta testing has been extremely appreciated.
